madara/crates/primitives/bloom_filter/src/filter.rs

```diff
@@ -0,0 +1,280 @@
+    pub fn finalize(self) -> BloomFilter<H, BitStore> {
+        BloomFilter {
+            storage: self.storage.to_readonly(),
+            bit_size: self.bit_size,
+            hash_count: self.hash_count,
+            _hasher: PhantomData,
+        }
+    }
+
+    /// Adds an item to the Bloom filter.
+    ///
+    /// # Arguments
+    ///
+    /// * `item`: The item to add, which must implement the Hash trait
+    pub fn add<T: Hash>(&self, item: &T) {
+        calculate_hashes::<H, T>(self.hash_count, item)
+            .for_each(|hash| self.storage.set_bit((hash % self.bit_size) as usize));
+    }
+}
+
+impl<H: Hasher + Default, B> BloomFilter<H, B> {
+    /// Returns the size of the Bloom filter in bits.
+    pub fn size(&self) -> u64 {
+        self.bit_size
+    }
+
+    /// Returns the number of hash functions used by the filter.
+    pub fn hash_count(&self) -> u8 {
+        self.hash_count
+    }
+
+    pub(crate) fn storage(&self) -> &B {
+        &self.storage
+    }
+}
+
+impl<H: Hasher + Default> BloomFilter<H, BitStore> {
+    /// Creates a new immutable Bloom filter from a bit storage backend.
+    pub fn from_storage(storage: BitStore, hash_count: u8) -> Self {
+        let bit_size = storage.len() as _;
+        Self { storage, bit_size, hash_count, _hasher: PhantomData }
+    }
+
+    /// Tests whether an item might be in the set (immutable version).
+    pub fn might_contain<T: Hash>(&self, item: &T) -> bool {
+        calculate_hashes::<H, T>(self.hash_count, item)
+            .all(|hash| self.storage.test_bit((hash % self.bit_size) as usize))
+    }
+
+    /// Tests whether an item might be in the set using pre-calculated hash values.
+    pub fn might_contain_hashes(&self, pre_calculated: &PreCalculatedHashes) -> bool {
+        pre_calculated.get_positions_for_size(self.bit_size).all(|pos| self.storage.test_bit(pos))
+    }
+
+    /// Estimates the current false positive rate of the filter based on
+    /// the number of bits set and the number of hash functions.
+    pub fn estimated_false_positive_rate(&self) -> f64 {
+        let set_bits = self.storage.count_ones() as f64;
+        let total_bits = self.storage.len() as f64;
+        (set_bits / total_bits).powi(self.hash_count as i32)
+    }
+}
+
+/// Calculates multiple hash positions for an item using double hashing technique.
+///
+/// This function implements a double hashing strategy to generate k different bit
+/// positions using the formula:
+///
+/// h(i) = (h1 + i * h2) mod m
+///
+/// where:
+/// - h1 is the primary hash value of the item
+/// - h2 is a secondary hash derived from h1 using rotation and golden ratio multiplication
+/// - m is the bit array size (applied implicitly by the caller)
+/// - i iterates from 0 to k-1, with k being the number of desired hash functions
+///
+/// # Mathematical Justification
+///
+/// ## Double Hashing
+/// - Double hashing is used to minimize clustering in Bloom filters.
```

Comment:
Okay, I was thinking that we don't need that because the keys are always uniform
But in fact they're not - the from_address might be (if we forget about contract address upper bound) but event keys are user-made and arbitrary, they also usually encode ascii
So you still need that - got it